arrows/ffmpeg/ffmpeg_video_input.cxx

```diff
@@ -1697,71 +1798,204 @@ ::seek( kv::frame_id_t frame_number )
       "Could not seek to frame ", frame_number );
     if( codec_context )
     {
       avcodec_flush_buffers( codec_context.get() );
     }
 
     // Move forward through frames until we get to the desired frame
+    size_t advance_count = 0;
     do
     {
-      advance();
-      if( at_eof )
+      advance( advance_count == 0 );
+      ++advance_count;
+      if( mode != SEEK_MODE_KEYFRAME_AFTER && at_eof )
       {
         throw_error(
           "Could not seek to frame ", frame_number + 1, ": "
           "End of file reached" );
       }
-    } while( this->frame_number() < frame_number );
+    } while( ( mode == SEEK_MODE_EXACT &&
+               this->frame_number() < frame_number ) ||
+             ( mode == SEEK_MODE_KEYFRAME_AFTER &&
+               ( ( this->frame_number() < frame_number ) ||
+                 ( !at_eof && !frame->frame->key_frame ) ) ) );
 
     // Check for success
-    if( this->frame_number() == frame_number )
+    if( ( mode == SEEK_MODE_EXACT && this->frame_number() == frame_number ) ||
+        ( mode != SEEK_MODE_EXACT && frame && frame->frame->key_frame &&
+          ( ( mode == SEEK_MODE_KEYFRAME_BEFORE )
+            ? this->frame_number() <= frame_number
+            : this->frame_number() >= frame_number ) ) ||
+        ( mode == SEEK_MODE_KEYFRAME_AFTER && at_eof ) )
     {
-      break;
+      if( parent->klv_enabled && advance_count <= 1 && false )
+      {
+        auto const chosen_frame_number = this->frame_number();
+        converted_timestamp = frame_number - backstep - 1;
+        converted_timestamp =
+        av_rescale_q(
+          converted_timestamp, av_inv_q( frame_rate() ),
+          video_stream->time_base );
+        converted_timestamp += start_ts;
+
+        clear_state_for_seek();
+        throw_error_code(
+          av_seek_frame(
+            format_context.get(), video_stream->index, converted_timestamp,
+            AVSEEK_FLAG_BACKWARD | AVSEEK_FLAG_ANY ),
+          "Could not seek to frame ", frame_number );
+        if( codec_context )
+        {
+          avcodec_flush_buffers( codec_context.get() );
+        }
+
+        advance_count = 0;
+        do
+        {
+          advance( advance_count == 0 );
+          ++advance_count;
+          if( at_eof || this->frame_number() > chosen_frame_number )
+          {
+            throw_error(
+              "Could not seek to frame ", frame_number + 1, ": "
+              "KLV re-seek failed" );
+          }
+        } while( this->frame_number() < chosen_frame_number );
+
+        if( advance_count <= 1 )
+        {
+          LOG_WARN(
+            parent->logger,
+            "KLV re-seek failed; "
+            "KLV reported for first frame may be incomplete" );
+        }
+      }
+
+      if( mode != SEEK_MODE_EXACT )
+      {
+        auto& image_packets = frame->get_raw_image().packets;
+        for( auto it = image_packets.begin(); it != image_packets.end(); )
```

Comment:
Why not to use while instead? Looks more like a while loop

Reply:
See response to similar suggestion above.